nixos/modules/config/sysfs.nix

```diff
@@ -0,0 +1,192 @@
+{
+  lib,
+  config,
+  ...
+}:
+
+let
+  inherit (builtins)
+    head
+    isAttrs
+    isFloat
+    isList
+    length
+    match
+    tail
+    ;
+
+  inherit (lib)
+    concatStringsSep
+    flatten
+    floatToString
+    foldl'
+    isDerivation
+    mapAttrsToList
+    nameValuePair
+    removePrefix
+    throwIf
+    ;
+
+  inherit (lib.options)
+    showDefs
+    showOption
+    ;
+
+  inherit (lib.strings)
+    isConvertibleWithToString
+    ;
+
+  inherit (lib.path.subpath) join;
+
+  cfg = config.boot.kernel.sysfs;
+
+  sysfsAttrs = with lib.types; nullOr (either sysfsValue (attrsOf sysfsAttrs));
+  sysfsValue = lib.mkOptionType {
+    name = "sysfs value";
+    description = "sysfs attribute value";
+    descriptionClass = "noun";
+    check = v: isConvertibleWithToString v && mkValueString v != "";
+    merge =
+      loc: defs:
+      if length defs == 1 then
+        (head defs).value
+      else
+        (foldl' (
+          first: def:
+          # merge definitions if they produce the same value string
+          throwIf (mkValueString first.value != mkValueString def.value)
+            "The option \"${showOption loc}\" has conflicting definition values:${
+              showDefs [
+                first
+                def
+              ]
+            }"
+            first
+        ) (head defs) (tail defs)).value;
+  };
+
+  mapAttrsToListRecursive =
+    fn: set:
+    let
+      recurse =
+        p: v:
+        if isAttrs v && !isDerivation v then mapAttrsToList (n: v: recurse (p ++ [ n ]) v) v else fn p v;
+    in
+    flatten (recurse [ ] set);
+
+  mkPath = p: "/sys" + removePrefix "." (join p);
+
+  mkValueString =
+    v:
+    # true will be converted to "1" by toString, saving one branch
+    if v == false then
+      "0"
+    else if isFloat v then
+      floatToString v # warn about loss of precision
+    else if isList v then
+      concatStringsSep "," (map mkValueString v)
+    else
+      toString v;
+
+  tmpfiles = builtins.listToAttrs (
+    mapAttrsToListRecursive (
+      p: v: if v == null then [ ] else nameValuePair (mkPath p) { w.argument = mkValueString v; }
+    ) cfg
+  );
+in
+{
+  options = {
+    boot.kernel.sysfs = lib.mkOption {
+      type = lib.types.submodule {
+        freeformType = lib.types.attrsOf sysfsAttrs // {
+          description = "nested attribute set of null or sysfs attribute values";
+        };
+      };
+
+      description = ''
```

Comment:
I am not sure if that will display properly on search.nxios.org because of its length.

Reply:
Is there a conventional upper limit on the description length?

I could perhaps elaborate on the option in a section of the NixOS manual and refer to it in this description.